### ISSUE TITLE
Build infra: Finish parallel builds by finalizing persistence across stages

### ISSUE THREAD

**Issue by pbakaus** [ampproject/docs#895]
The work on the branch https://github.com/ampproject/docs/tree/parallel parallelizes the grow build by splitting up generation across locales, then later merging the results and deploying them once as final step.

Unfortunately, Travis build steps do not have any data sharing/persistence, so we need to upload results from individual steps to a secondary service, then download them in the final step, then deploy to firebase.

In the current version, I'm trying to use Google Cloud Storage, as our production and master firebase instances already come with a bucket. Setting up all gcloud on travis is unfortunately a lot of overhead, but seems to be the way to go to authenticate. Instead, I tried to install [gsutil](https://cloud.google.com/storage/docs/gsutil) independently, but gsutil does not have a non-interactive way to easily authenticate as service account (argh). For testing purposes, I set both buckets to `public-read-write` (ACL), which is obviously not a good solution, and also doesn't seem to work with the `gsutil rsync` command.

**Comment by bpaduch** [comment]
Would love to see this implemented .... soon!

**Comment by rsimha** [comment]
@pbakaus @bpaduch I just picked this up. Before working on it, I had to rebase `docs/parallel` on `docs/master`. (The new commit is at https://github.com/ampproject/docs/commit/1177b5c9518e59d63bd48c2a56ae19fc70dc38ca)

Next, I'm watching the resulting Travis build at https://travis-ci.org/ampproject/docs/builds/371154796 to see what's missing.

Once we have results, I'll likely need some input from you on what's missing, and what the likely solution might be.

**Comment by rsimha** [comment]
From the build logs, it seems like the default version of python on Travis is incompatible.

```
$ pip install gsutil --user
Collecting gsutil
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:318: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
```

Trying a newer version.

**Comment by rsimha** [comment]
@pbakaus Still seeing errors, but the logs are easier to read: https://travis-ci.org/ampproject/docs/jobs/371219968 (You'll need to expand the various sections to see the errors)

I can dig around this some more, but it would help if you can point me to what you think ought to be done here.

**Comment by rsimha** [comment]
@pbakaus Have you seen https://docs.travis-ci.com/user/deployment/gcs/ ? I wonder if we can use that setup to upload the build output from each parallel job.

**Comment by rsimha** [comment]
I think https://github.com/GoogleCloudPlatform/gsutil/issues/457#issuecomment-326409413 contains a clue to why the `gsutil` operation in the `Deploy` stage is [failing](https://travis-ci.org/ampproject/docs/jobs/377026178#L680). Note that the `Build` stage is [successful](https://travis-ci.org/ampproject/docs/jobs/377026177#L3364).

I suspect that the anonymous `gsutil` user that uploads the contents of the build directory in the parallel `Build` stages is not the same as the anonymous user that tries to download them in the `Deploy` stage.

I'd imagine we can solve this by authenticating. We can try to get this to work using the link in my comment above this one.

**Comment by pbakaus** [comment]
Thanks Raghu for your research! I've did as you suggested, but for some reason other than that, two of the parallel builds are currently failing for no apparent reason I can see in the log...any idea?

**Comment by rsimha** [comment]
@pbakaus Those errors happen during `grow build`, and seem unrelated to the upload.

If you've enabled anonymous download from the GCE instance, let me try rebasing and kicking off a new `ampproject/docs/parallel` build.

**Comment by rsimha** [comment]
**Update:** Good news. I was able to successfully download the previous build with this:
```
gsutil -m rsync -r gs://ampproject-staging.appspot.com/ build
```
I expect this will now work from the `Deploy` stage on Travis. Let's try it.

I've kicked off a new build at https://travis-ci.org/ampproject/docs/builds/377958267

**Comment by rsimha** [comment]
@pbakaus Parallel builds are now working, but they're flaky due to #968.

**Comment by rsimha** [comment]
@pbakaus This worked for a while, and we're now running into quota errors on the app engine instance:
```
Copying file://build_None_en_de_es.zip [Content-Type=application/zip]...
ResumableUploadAbortException: 403 The App Engine application does not have enough quota.
CommandException: 1 file/object could not be transferred.
```

**Comment by pbakaus** [comment]
I'm not actually sure how to fix it as Cloud Console is too confusing for me, but I gave you admin rights to the staging and production projects, maybe you'll find it..

**Comment by rsimha** [comment]
@pbakaus Billing hadn't been enabled for the staging project. I've enabled it, and things seem to be working again.

**Comment by rsimha** [comment]
This is now working: https://travis-ci.org/ampproject/docs/builds/385647290

**Comment by bpaduch** [comment]
@rsimha - consider closing?

**Comment by rsimha** [comment]
Wasn't this closed by bb2be98?

**Comment by bpaduch** [comment]
Yeah.  Sorry. Wrong bug I was looking at.
